### PR TITLE
DOC: Document code contributors on website

### PR DESCRIPTION
**Pull request by larsoner**
This gets us started on keeping track of code contributions to MNE, the first part of #11605.

1. Create a "Contributors" page, and link to it from the main landing page
2. Add a "Code Credit" heading on that page, mentioning eventually we'll add other stuff.
3. Populate it with code credit based on lines changed in commits in MNE-Python:

   1. Overall changes to MNE-Python
   2. submodule-by-submodule credit

We have *a lot* of contributors and *a lot* of modules, so it's tough to get all the information in there. There are lots of potential ways to do this, what to include, how to show it -- and thus a lot of potential bike shedding, so we should probably just try to make sure what's here is an improvement over the author lists we currently use. Let's not shoot for perfection here but just "good enough"!

Still some stuff to do:

- [x] Manually run and create the page
- [x] Converge on the idea it's actually better
- [x] Make the page rebuild, likely by running on CircleCI (or opt-in locally), perhaps in a directive rather than writing an RST file
- [x] Ensure that the line counts are actually correct / reasonable, considering the squash+merge (new) vs standard merge (old) ways of doing things. I think it's probably not correct yet, but I'll think on it. (I started writing something with the `gh` command-line tool, but querying every merged PR and get files changed and lines changed in each file takes forever. Getting the merge commit -- so that we could then use the `git history` to figure things out -- sometimes returns commit hashes that don't exist in our codebase. So it's not trivial.)
- [x] Get reviews
- [x] Unzip `prs.zip`
- [ ] Merge
- [ ] Remove author names from `mne/` files in separate PR

### CODE REVIEW

**Comment by larsoner** [comment]
Frontpage

Screenshot 2024-08-02 at 10 16 37 PM

Links to

Screenshot 2024-08-02 at 10 18 22 PM

https://output.circle-artifacts.com/output/job/1d4da1a2-6f08-4643-b1f9-16582ba5a3df/artifacts/0/html/credit.html

**Comment by agramfort** [comment]
looks nice !

File: tools/dev/credit.py
Comment (sappelhoff) [inline tools/dev/credit.py:21]: really nice work here, thanks!

would it make sense to use https://github.com/mne-tools/mne-python/blob/main/.git-blame-ignore-revs for this?

Comment (larsoner) [reply tools/dev/credit.py:21]: Yes in principle! I actually do use the `.git-blame-revs-ignore` in the `.sh` file. There is probably a way to map the commit numbers in the ignore file back to the PRs they came from, which would allow this script to do some similar ignoring. I'll look into it if I come back to using this script -- right now everything is done in the bash script because doing it in Python is too slow!

File: .github/workflows/credit.yml
Comment (larsoner) [inline .github/workflows/credit.yml:42]: This seems like it should in principle work on `main`. I on purpose left out our most recently merged PR #12747 and it got to the stage it would try to open a PR to add it 

https://github.com/mne-tools/mne-python/actions/runs/10275063578/job/28433018563?pr=12774

I set it to try nightly for now. Once we know it works I plan to change it to monthly (seems often enough).

File: doc/credit.rst
Comment (larsoner) [inline doc/credit.rst:2]: This file is autogenerated, better to review it by looking at the rendered HTML and code that generates it.

**Comment by larsoner** [comment]
Okay @drammock I think I'm ready for review here!

What I came up with was:

1. Having separate `doc/sphinxext/prs/<pr_num>.json` files, one per PR, with the info we need to come up with code credit. Eventually if we want we can add other stuff like PR comment counts to those, and maybe issue comment counts in `issues/<issue_num>.json` or whatever. So I think the format is pretty flexible.
2. Getting the change info is done using PyGithub in `tools/dev/update_credit_json.py` so we stay Pythonic. Only downside is that the initial gathering of data is quite slow. But incremental monthly updates should be fast, and I've added a GitHub action that will eventually run monthly to do it for us. This ultimately seemed preferable over using, say, `git diff` locally because PyGithub gives us easier access to more info and future extensibility, and also the files-changed-in-PR is something we get directly there, and something that was really quite difficult using `git` locally (for me at least!). 
3. Mapping the change counts to a `rst` file is done in `tools/dev/update_credit_rst.py`, and consumes the `.json` files. This does all sorts of accounting stuff, like mapping between author names, grouping into module types, etc. And then it takes that info and decides how to present it, which is as author badges embeddeding in module cards. Lots of opinionated choices here that we can tweak (maybe mostly in follow-up PRs?) but it's hopefully better than what we have in `main` already.

For ease of reviewing, I zipped the json files to a single zip, as I assumed that without doing so the `Files` page on this PR was going to be unusable. So once we agree this is ready for merge, I'll go back to individual `.json` files, make sure it still looks okay, and merge.

**Review by drammock — COMMENTED** [review]
did interactive review with @larsoner  in real-time, hence no in-line suggestions. Approving on assumption that he took good notes and will fix everything we discussed. 

These are explicitly NOT in scope for this PR (but can be in follow-up PRs):

- tweaks to visual presentation
- changes to which modules we show credit for
- adding additional metrics (such as PR reviews) --- again, we want to do this, and can/should do so in follow-up PRs.
- performance enhancements relating to the GH API queries / JSON parsing (I'm convinced that it's fast enough for now)

**Review by drammock — APPROVED** [review]

**Comment by britta-wstnr** [comment]
Nice, I like it!
One nitpick - if I convert to greyscale, the two colors for the top 10% contributors and rest are not distinguishable anymore:
image
Can we pick colors that are more easily distinguishable (assuming this would be an easy fix)?

**Comment by larsoner** [comment]
Okay changed to something with more contrast that looks okay in grayscale, will merge once I manually check by eye it rendered correctly!

File: doc/conf.py
Comment (larsoner) [inline doc/conf.py:1660]: FYI @drammock I ended up changing the structure a bit. I realized that having a `credit.rst` in our code history that gets updated with thousands of badges is less than ideal -- it'll change a lot and bloat things I think. So instead now I have it create a `doc/code_credit.inc` on the fly (it's very fast) at the start of the doc build so that we don't have to keep it in our history. In theory this could instead be a directive, but I think it's actually nicer as a `.inc` because you can inspect it after it's generated, don't have to deal with docutils and nested parsing at all, etc.

**Pull request merged**
